File: middleware/RAGMgmt-Service/common/corpus_types.py
"""Three-corpora taxonomy for the Project A grounding model.

The README commits to grounding every draft in three corpora:
  1. Regulator's published advertising rules
  2. The practice's own voice corpus (tone/brand reference)
  3. Open-access clinical evidence

This module is the source of truth for which seeded `dataset_name` belongs to
which corpus. It stays in sync by hand with:
  * DataMgmt-Service/migrations/rag_app/V004__seed_initial_endpoints_and_datasets.sql
    (the original 5 datasets), and
  * V009__seed_practice_voice_dataset.sql (Practice_Voice_Sample).

If/when a cross-DB lookup or a denormalized `dataset_type` column on
`child_chunk_embeddings` lands, this constant becomes a fallback rather than the
authoritative map. Until then, adding a new seeded dataset means updating both
the migration AND this module.
"""

from __future__ import annotations

REGULATOR = "regulator"
CLINICAL_EVIDENCE = "clinical_evidence"
PRACTICE_VOICE = "practice_voice"

ALL_CORPUS_TYPES: tuple[str, ...] = (REGULATOR, CLINICAL_EVIDENCE, PRACTICE_VOICE)


DATASET_NAME_TO_CORPUS_TYPE: dict[str, str] = {
    # Regulator advertising rules.
    "Public_Regulator_Guidelines":   REGULATOR,
    # Open-access clinical evidence — every dataset on the evidence side.
    "NCBI_PubMed":                   CLINICAL_EVIDENCE,
    "PMC_Open_Access_Subset":        CLINICAL_EVIDENCE,
    "Kaggle_Medical_Transcriptions": CLINICAL_EVIDENCE,
    "Common_Crawl_Allied_Health":    CLINICAL_EVIDENCE,
    # Practice voice / brand reference.
    "Practice_Voice_Sample":         PRACTICE_VOICE,
}


def datasets_for_corpus(corpus_type: str) -> list[str]:
    """Return the dataset_names that belong to a given corpus, in stable order."""
    return [d for d, c in DATASET_NAME_TO_CORPUS_TYPE.items() if c == corpus_type]


def corpus_for_dataset(dataset_name: str) -> str | None:
    return DATASET_NAME_TO_CORPUS_TYPE.get(dataset_name)

File: portals/customer/src/app/core/corpus-types.ts
/**
 * Three-corpora taxonomy — mirrors RAGMgmt-Service/common/corpus_types.py and
 * the admin portal's copy at portals/admin/src/app/core/corpus-types.ts.
 *
 * Kept frontend-local rather than fetched from the API so the catalog and
 * generate screens can render corpus chips on every dataset/citation without
 * an extra round-trip on every page load.
 *
 * When a new dataset_name is added by a backend migration, both this module
 * AND the Python equivalent need an entry. The Python-side test
 * `test_seeded_datasets_present` enforces alignment on the backend; this
 * module is the corresponding frontend source of truth.
 */

import { CorpusType } from './api/ragmgmt-api.service';

export const CORPUS_LABEL: Record<CorpusType, string> = {
  regulator: 'Regulator',
  clinical_evidence: 'Clinical evidence',
  practice_voice: 'Practice voice',
};

const DATASET_NAME_TO_CORPUS_TYPE: Record<string, CorpusType> = {
  // Regulator advertising rules.
  Public_Regulator_Guidelines: 'regulator',
  // Open-access clinical evidence.
  NCBI_PubMed: 'clinical_evidence',
  NCBI_PubMed_Abstracts: 'clinical_evidence',
  PMC_Open_Access_Subset: 'clinical_evidence',
  PMC_Open_Access_FullText: 'clinical_evidence',
  Kaggle_Medical_Transcriptions: 'clinical_evidence',
  Common_Crawl_Allied_Health: 'clinical_evidence',
  // Practice voice / brand reference.
  Practice_Voice_Sample: 'practice_voice',
};

export function corpusForDataset(datasetName: string | null | undefined): CorpusType | null {
  if (!datasetName) return null;
  return DATASET_NAME_TO_CORPUS_TYPE[datasetName] ?? null;
}

export function corpusChipClass(corpus: CorpusType | null | undefined): string {
  if (!corpus) return 'corpus-chip corpus-unknown';
  return `corpus-chip corpus-${corpus.replace('_', '-')}`;
}

export function corpusLabel(corpus: CorpusType | null | undefined): string {
  if (!corpus) return 'unmapped';
  return CORPUS_LABEL[corpus];
}
